modules/hostdb/scan.go

```diff
@@ -79,10 +95,60 @@ func (hdb *HostDB) threadedProbeHost(entry *modules.HostEntry) {
 // every few hours to see who is online and available for uploading.
 func (hdb *HostDB) threadedScan() {
 	for {
+		// Determine who to scan. At most 'MaxActiveHosts' will be scanned,
+		// starting with the active hosts followed by a random selection of the
+		// inactive hosts.
+		id := hdb.mu.Lock()
+		{
+			// Check all of the active hosts.
+			for _, host := range hdb.activeHosts {
+				go hdb.threadedProbeHost(host.hostEntry)
+			}
+
+			// Assemble all of the inactive hosts into a single array and
+			// shuffle it.
+			i := 0
+			random := make([]*hostEntry, len(hdb.allHosts))
+			for _, entry := range hdb.allHosts {
+				random[i] = &entry
+				i++
+			}
+
+			// Randomize the slice by swapping each element with an element
+			// that hasn't been visited yet.
+			for i := 0; i < len(hdb.allHosts); i++ {
+				N, err := rand.Int(rand.Reader, big.NewInt(int64(len(hdb.allHosts)-i)))
+				if err != nil {
+					if build.DEBUG {
+						panic(err)
+					}
+				} else {
+					break
+				}
+
+				n := int(N.Int64()) + i
+				tmp := random[i]
+				random[i] = random[n]
+				random[n] = tmp
+			}
+
+			// Select the first InactiveHostCheckupQuantity hosts from the
+			// shuffled list.
```

Comment:
seems like overkill, but I may be wrong. The important factors are: `InactiveHostCheckupQuantity`, `len(allHosts)`, and how frequently `Scan` is run. If probing inactive hosts is low-priority, then it would probably suffice to just select N random hosts (without the guarantee of uniqueness).